core/src/com/etheller/warsmash/parsers/jass/Jass2.java

```diff
@@ -861,380 +695,341 @@ public JassValue call(final List<JassValue> arguments, final GlobalScope globalS
-						if (target != null) {
-							if (attackType == null) {
-								attackType = CAttackType.UNKNOWN;
-							}
-
-							if (weaponType == null) {
-								weaponType = CWeaponSoundTypeJass.WHOKNOWS;
+			jassProgramVisitor.getJassNativeManager().createNative("GroupEnumUnitsOfTypeCounted", (arguments, globalScope, triggerScope) -> {
+				final List<CUnit> group = arguments.get(0).visit(ObjectJassValueVisitor.<List<CUnit>>getInstance());
+				final String unitname = arguments.get(1).visit(StringJassValueVisitor.getInstance());
+				final TriggerBooleanExpression filter = arguments.get(2).visit(ObjectJassValueVisitor.<TriggerBooleanExpression>getInstance());
+				final Integer countLimit = arguments.get(3).visit(IntegerJassValueVisitor.getInstance());
+				int count = 0;
+				for (final CUnit unit : CommonEnvironment.this.simulation.getUnits()) {
+					if (unitname.equals(unit.getUnitType().getLegacyName())) {
+						if (filter.evaluate(globalScope,
+								CommonTriggerExecutionScope.filterScope(triggerScope, unit))) {
+							// TODO the trigger scope for evaluation here might need to be a clean one?
+							group.add(unit);
+							count++;
+							if (count >= countLimit) {
+								break;
 							}
-
-							target.damage(CommonEnvironment.this.simulation, whichUnit, attackType, weaponType.name(),
-									(float) amount);
 						}
 					}
-
-					return BooleanJassValue.TRUE;
 				}
+				return null;
 			});
-			jassProgramVisitor.getJassNativeManager().createNative("GroupEnumUnitsOfTypeCounted", new JassFunction() {
-				@Override
-				public JassValue call(final List<JassValue> arguments, final GlobalScope globalScope,
-						final TriggerExecutionScope triggerScope) {
-					final List<CUnit> group = arguments.get(0).visit(ObjectJassValueVisitor.<List<CUnit>>getInstance());
-					final String unitname = arguments.get(1).visit(StringJassValueVisitor.getInstance());
-					final TriggerBooleanExpression filter = arguments.get(2)
-							.visit(ObjectJassValueVisitor.<TriggerBooleanExpression>getInstance());
-					final Integer countLimit = arguments.get(3).visit(IntegerJassValueVisitor.getInstance());
-					int count = 0;
-					for (final CUnit unit : CommonEnvironment.this.simulation.getUnits()) {
-						if (unitname.equals(unit.getUnitType().getLegacyName())) {
-							if (filter.evaluate(globalScope,
-									CommonTriggerExecutionScope.filterScope(triggerScope, unit))) {
+			jassProgramVisitor.getJassNativeManager().createNative("GroupEnumUnitsInRect", (arguments, globalScope, triggerScope) -> {
+				final List<CUnit> group = arguments.get(0).visit(ObjectJassValueVisitor.<List<CUnit>>getInstance());
+				final Rectangle rect = arguments.get(1).visit(ObjectJassValueVisitor.<Rectangle>getInstance());
+				final TriggerBooleanExpression filter = nullable(arguments, 2, ObjectJassValueVisitor.<TriggerBooleanExpression>getInstance());
+				CommonEnvironment.this.simulation.getWorldCollision().enumBuildingsInRect(rect, (unit) -> {
```

Comment:
Through some human error, this function call has mysteriously changed from `enumUnitsInRect` to `enumBuildingsInRect` in your changes. I imagine this may have been a matter of manually typing the function call with an autosuggest feature.

Conceptually, when enumerating all units in a rect, the user will want to consider units of any kind, regardless of whether or not the unit is a structure.